### PR TITLE
CRSF modules baudrate

### PR DESCRIPTION
**Pull request by 3djc**
This rationalize CRSF modules baurate.

Internal module are defined in radio hardware
image

On T-Pro the baudrate value is limited to 1.8Mbps given there is no DMA possible

External module are defined at model level (since you can have a model using a Crossfire module and another using a ELRS module, both running at different speed)
image

Note : baudrate are displayed only for CRSF type module since they are the only one not fixed

Note on GHOST modules : 
On radio that do have slow inverters (x7,..), a limited choice of 115k/400 is proposed to allow people who have modded their radio to get 400k symetrical

~~PLEASE NOTE THAT THE COMPANION SIDE NEEDS TO BE DONE~~

### CODE REVIEW

File: radio/src/datastructs_private.h [hunk 1]
```diff
@@ -870,7 +874,7 @@ PACK(struct RadioData {
   int8_t timezone:5;
   uint8_t adjustRTC:1;
   NOBACKUP(uint8_t inactivityTimer);
-  uint8_t telemetryBaudrate:3;
+  uint8_t InternalModuleBaudrate:3;
```
Comment (elecpower) [inline radio/src/datastructs_private.h:877]: Camel case

File: radio/src/storage/yaml/yaml_datastructs_nv14.cpp [hunk 2]
```diff
@@ -389,7 +389,7 @@ static const struct YamlNode struct_RadioData[] = {
   YAML_SIGNED( "timezone", 5 ),
   YAML_UNSIGNED( "adjustRTC", 1 ),
   YAML_UNSIGNED( "inactivityTimer", 8 ),
-  YAML_UNSIGNED( "telemetryBaudrate", 3 ),
+  YAML_UNSIGNED( "InternalModuleBaudrate", 3 ),
```
Comment (elecpower) [inline radio/src/storage/yaml/yaml_datastructs_nv14.cpp:392]: Camel case

**Comment by CapnBry** [comment]
Wowie I was just talking with someone today and they had to run 400k for everything because they had an external Crossfire. Not gonna be a problem any more!

File: radio/src/storage/yaml/yaml_datastructs_nv14.cpp [hunk 3]
```diff
@@ -247,9 +247,9 @@ const struct YamlIdStr enum_PotsWarnMode[] = {
   {  0, NULL  }
 };
 const struct YamlIdStr enum_ModelOverridableEnable[] = {
-  {  OVERRIDE_VALUE_GLOBAL, "GLOBAL"  },
-  {  OVERRIDE_VALUE_OFF, "OFF"  },
-  {  OVERRIDE_VALUE_ON, "ON"  },
+  {  OVERRIDE_VALUE_GLOBAL, "VALUE_GLOBAL"  },
+  {  OVERRIDE_VALUE_OFF, "VALUE_OFF"  },
+  {  OVERRIDE_VALUE_ON, "VALUE_ON"  },
```
Comment (elecpower) [inline radio/src/storage/yaml/yaml_datastructs_nv14.cpp:252]: Why the need to add VALUE_ prefix for yaml?

**Comment by elecpower** [comment]
As you are changing the data struct should you not have data conversion for existing models?

File: radio/src/storage/yaml/yaml_datastructs_nv14.cpp (hunk 3 above)
Comment (3djc) [reply radio/src/storage/yaml/yaml_datastructs_nv14.cpp:252]: Unsure, that got genrated from the yaml generator

**Comment by 3djc** [comment]
I think it is ok like that because default are ok (they will work in any case), then user can optimize given their need, which we cannot really extrapolate from their current settings

File: radio/src/storage/yaml/yaml_datastructs_t12.cpp [hunk 4]
```diff
@@ -237,9 +237,9 @@ const struct YamlIdStr enum_PotsWarnMode[] = {
   {  0, NULL  }
 };
 const struct YamlIdStr enum_ModelOverridableEnable[] = {
-  {  OVERRIDE_VALUE_GLOBAL, "GLOBAL"  },
-  {  OVERRIDE_VALUE_OFF, "OFF"  },
-  {  OVERRIDE_VALUE_ON, "ON"  },
+  {  OVERRIDE_VALUE_GLOBAL, "VALUE_GLOBAL"  },
+  {  OVERRIDE_VALUE_OFF, "VALUE_OFF"  },
+  {  OVERRIDE_VALUE_ON, "VALUE_ON"  },
```
[the remark this answers is not in the document]
Comment (raphaelcoeffic) [reply radio/src/storage/yaml/yaml_datastructs_t12.cpp:242]: @pfeerick this has been messed up in #1363, as it seems that the changes to YAML files have been done by hand and not using the generator.

**Comment by 3djc** [comment]
I'm thinking about adding an extra status line, that would display something like 
Status     500Hz, 0 errs

**Comment by CapnBry** [comment]
I added my own line to the sync status that's enabled with debug (and added it to Crossfire type modules) to be able to help visualize `G`ood,  `C`RC failed, and `F`raming errors (incremented when the first byte received by the crsf processor when the first byte of a packet is not the sync byte). The values were tallied per second.
sync-status

It was specific to CRSF protocol only, but I could see a use for somehow showing the user how many errors they're getting. I think the order of importance of stats goes something like this:
* Hardware error count
* Mixer rate
* Protocol-level good telemetry packets received
* Protocol-level error telemetry packets received (including failed sync, length failed sanity check, buffer full too, or these could be split out possibly if this becomes a whole screen)
* Mixer sync offset

I think you've got the best two already. If it is possible to put a long press handler on the status item to reset the error count it would be cool too, since I see more long term benefit to having an accumulated total number of errors instead of per-second since usually errors are spread out.

**Comment by 3djc** [comment]
I was thinking something useful for end users, not as sophisticated as yours, but surely the reset there would be a nice touch

**Comment by CapnBry** [comment]
Yeah I agree it is too Inside Baseball to show all that to the users, but I do think being able to see the error count would be very useful, since ELRS doesn't have any way of knowing how many telemetry packets are actually being received by the handset properly. Being able to see that the errors goes up right at the time they get the "Telemetry Lost/Telemetry Recovered" callout would be a clear indicator of what caused it.

I don't think anything else on my list is in the same tier of usefulness as just the simple hardware error count, given that ELRS generally doesn't generate malformed CRSF packets (except for when I am at the wheel!).

**Comment by 3djc** [comment]
image

**Comment by CapnBry** [comment]
Looks good to me and how do we get 12bit raw support too gimme gimme. (Although I know it would be best implementing `CRSF_FRAMETYPE_SUBSET_RC_CHANNELS_PACKED` and allow 10-13 bit :-D)

File: radio/src/storage/yaml/yaml_datastructs_t12.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (pfeerick) [reply radio/src/storage/yaml/yaml_datastructs_t12.cpp:242]: Yup, as mentioned here ;) https://github.com/EdgeTX/edgetx/pull/1581#discussion_r805139974

[the remark this answers is not in the document]
Comment (pfeerick) [reply radio/src/storage/yaml/yaml_datastructs_t12.cpp:242]: Should be resolved through #1586

**Comment by 3djc** [comment]
image

Colorlcd works too

**Comment by 3djc** [comment]
Latest version
image

**Comment by pfeerick** [comment]
External RF looks good on TX16S:
image

For some weird reason, first the set value, and then remainder of array is being shown when Internal is set to either 1.87M or 3.75M... other values all seem fine though:
image
image

Noticed this first with a rebased version of PR, and then with PR as is. Present on both simu and hardware. 
Next up with be B&W - Zorro or TPro :)

**Comment by raphaelcoeffic** [comment]
I suggest rebasing on top of #1586 once it has been merged.

**Comment by pfeerick** [comment]
@raphaelcoeffic Well, other than the minor issue that compaion wasn't reading the current baud rate value, this seems to be working fine. I have made a change that makes it read and convert the value, but perhaps it could be done better.

Two questions:
1) Is showing of baud rate option for internal as well as external RF a precursor to your serial refactor PR, or an oversight?

2) The TPro is limited to 1.87M for internal baud... does limit apply to the external RF as well? If so, may firmware and companion need to be limited for EXT RF also? 

image

**Comment by raphaelcoeffic** [comment]
> I have made a change that makes it read and convert the value, but perhaps it could be done better (i.e. I'm not even sure if it is correctly making the connection to the current board variant, but it gives the right values anyway)

Looks fine to me. I think it applies to Ghost as well, however.

**Comment by raphaelcoeffic** [comment]
> Is showing of baud rate option for internal as well as external RF a precursor to your serial refactor PR, or an oversight?

*oversight*, it should be shown only for external, as internal is set somewhere else.

**Comment by raphaelcoeffic** [comment]
> The TPro is limited to 1.87M for internal baud... does limit apply to the external RF as well? If so, may firmware and companion need to be limited for EXT RF also?

I don't think it is limited on the external interface. At least not specifically, as all radios have the same capabilities there in terms of sending DMA vs. IRQ based.

**Comment by pfeerick** [comment]
Perfect. I can deal with both of those then I think. It was only going to be the baud rate limit that was going to be a sticking point for me ;) 

I've also realised I don't fully understand the byte alignment part of https://github.com/EdgeTX/edgetx/pull/1586#issuecomment-1068050050 so don't know how to proceed beyond this (and then re-generating the yaml structs, of course)...


```diff
diff --git a/radio/src/datastructs_private.h b/radio/src/datastructs_private.h
index a69b9c0a1..5a6803c41 100644
--- a/radio/src/datastructs_private.h
+++ b/radio/src/datastructs_private.h
@@ -872,6 +872,7 @@ PACK(struct RadioData {
   int8_t timezone:5;
   uint8_t adjustRTC:1;
   NOBACKUP(uint8_t inactivityTimer);
+  CUST_ATTR(telemetryBaudrate, r_telemetryBaudrate, nullptr);
   uint8_t internalModuleBaudrate:3;
   SPLASH_MODE; /* 3bits */
   int8_t hapticMode:2 CUST(r_beeperMode,w_beeperMode);
diff --git a/radio/src/storage/yaml/yaml_datastructs_funcs.cpp b/radio/src/storage/yaml/yaml_datastructs_funcs.cpp
index 50849ac07..149998780 100644
--- a/radio/src/storage/yaml/yaml_datastructs_funcs.cpp
+++ b/radio/src/storage/yaml/yaml_datastructs_funcs.cpp
@@ -1727,3 +1727,10 @@ static void r_jitterFilter(void* user, uint8_t* data, uint32_t bitoffs,
   uint32_t i = yaml_str2uint(val, val_len);
   yaml_put_bits(data, i, bitoffs, 1);
 }
+
+static void r_telemetryBaudrate(void* user, uint8_t* data, uint32_t bitoffs,
+                           const char* val, uint8_t val_len)
+{
+  uint32_t i = yaml_str2uint(val, val_len);
+  yaml_put_bits(data, i, bitoffs, 3);
+}

```

**Comment by pfeerick** [comment]
Right, so that should now read ghost telemetrybaudrate value also, and hides the combobox if the module being configured is the internal one. But the node will still be written to file... since encoding is oblivious to that :grin:

**Pull request merged**